src/refactor_nrepl/artifacts.clj

```diff
@@ -128,19 +138,88 @@
     (update-artifact-cache!))
   (->> @artifacts keys list*))
 
+(defn- artifact-versions* [artifact-id]
+  (->> (or (get @artifacts artifact-id)
+           (seq (get-mvn-versions! artifact-id))
+           (get-clojars-versions! artifact-id))
+       distinct
+       versions/version-sort
+       reverse
+       list*))
+
 (defn artifact-versions
   "Returns a sorted list of artifact version strings. The list can either come
   from the artifacts cache, the maven search api or the clojars search api in
   that order."
   [{:keys [artifact]}]
-  (->> (or (get @artifacts artifact)
-           (seq (get-mvn-versions! artifact))
-           (get-clojars-versions! artifact))
-       distinct
-       versions/version-sort
-       reverse
-       list*))
+  (artifact-versions* artifact))
+
+(defn- jar-at-the-top-of-dependency-hierarchy [new-deps]
+  ;; We only need to consider the dep at the top of the hierarchy because when we
+  ;; require those namespaces the rest of  the transitive deps will get pulled in
+  ;; too.
+  (letfn [(->jar [^File f]
+            (JarFile. f))]
+    (let [top-level-dep  (-> new-deps keys first)]
+      (-> top-level-dep meta :file ->jar))))
+
+(defn- make-resolve-missing-aware-of-new-deps!
+  "Once the deps are available on cp we still have to load them and
+  reset slamhound's cache to make resolve-missing work."
+  [^JarFile jar]
+  (doseq [new-namespace (find/find-namespaces-in-jarfile jar)]
+    (try
+      (require new-namespace)
+      (catch Exception _
+        ;; I've seen this happen after adding core.async as a dependency.
+        ;; It also happens if you try to require namespaces that no longer work,
+        ;; like compojure.handler.
+        ;; A failure here isn't a big deal, it only means that resolve-missing
+        ;; isn't going to work until the namespace has been loaded manually.
+        )))
+  (slamhound/reset)
+  (slamhound-regrow/clear-cache!))
+
+(defn- parse-coordinates [coordinates-str]
+  (let [coords (try (->> coordinates-str edn/read-string (take 2) vec)
+                    (catch Exception _))]
+    (if (and (= (count coords) 2)
+             (symbol? (first coords))
+             (string? (second coords)))
```

Comment:
Similarly, will this constraint work in face of :managed-dependencies?

With :managed-dependencies, a given member of :dependencies can be just `[foo/bar]` with no version specified.

Reply:
You'd call the op on the `:managed-dependencies` dep, and not on the `:dependencies` dep, where a version is available.

FWIW most people won't be calling hotload-deps directly on a dependency vector.  I think 99% of it's usage happens in the context of `cljr-add-project-dependency` where they've done a completed reading on artifact and then artifact version and cljr itself constructs a dependency vector to send down to refactor-nrepl

Reply:
> FWIW most people won't be calling hotload-deps directly on a dependency vector. 

Wouldn't it be kind of a usual case to make a desired change over project.clj, save the file (simply so that that work isn't lost and is reused in subsequent repls) and hit a single shortcut that would load the dep based on `POINT`?

Reply:
> make a desired change over project.clj, save the file 

`cljr-add-project-dep` does his part for you.  

I agree that we should have better support for calling `hotload-dependency` directly on a dep vector that you might have copied from elsewhere (which would then include classifiers and exclusions).

Perhaps we can just drop the validation and make sure the error message `pomegranate` throws surfaces to the user and is understandable?

Reply:
Yes, SGTM :)

Might be good to have some coverage for exclusions/classifiers anyway